### ISSUE TITLE
Add support for Code Inspection and Duplicate Finder results using NDepend

### ISSUE THREAD

**Issue by netwolfuk** [tcplugins/tcPrettyEmail#1]
I want to know if there is a way to add info on Code Inspection and Duplicate Finder results to the email that is sent out?

**Comment by sven-weber** [comment]
We are also looking for a way to include Code Inspection results. 
Are there any variables available that allow to adjust the templates in order to display this information?

**Comment by netwolfuk** [comment]
Hi @sven-weber

I never got to implement this as I couldn't come up with a generic solution which would work with all coverage tools.

Thinking about it now it seems like there are three approaches:
1. If the build has TeamCity coverage enabled it should be possible to get the coverage metrics from TeamCity and make them available to the velocity template.
2. Look for a published artifact with a specific name in a specific format. Then read the values from it and make them available to the template.
3. Look for a published artifact with a specific name and try to include it in the email body. I'm assuming that file is probably html, and the code would also need to find any images and attach them to the email too.

Do you think on of those solutions would work for you?

**Comment by sven-weber** [comment]
Hi @netwolfuk, 

thanks for your quick response!

In my opinion it would be the best to aim for solution one. 

TeamCity has a separate tab in the Build where the results of the Code Inspection are displayed. 
This tab seems to be already generic as it is displayed for all different inspection tools I have tested (Resharper, FxCop). 
![image](https://user-images.githubusercontent.com/39764491/53320920-5b284280-38d7-11e9-823d-16829957561c.png)
Maybe we can get the information we need from this tab? 

I also took at look at the plugin code and the regarding TeamCity documentations to see if they deliver the needed information.
The SRunningBuild object provided to the methods of the notifier has a method called getStaticValues. 
(see [here](http://javadoc.jetbrains.net/teamcity/openapi/9.1/jetbrains/buildServer/serverSide/SBuild.html#getBuildStatistics(jetbrains.buildServer.serverSide.BuildStatisticsOptions))) maybe these values contain a count of the code inspection errors. 
The BuildStatistics object, which can also be requested from the SRunningBuild, does unfortunately not deliver informations about the code inspection. 

I submitted a request about the problem to jetbrains (see [here](https://teamcity-support.jetbrains.com/hc/en-us/requests/1987900) and they suggested to include the count in the email title. Maybe that would also be appropriate? 

Thanks for you help,
I really appreciate it.

**Comment by netwolfuk** [comment]
HI @sven-weber.
Great find! I will have a look at the results returned by `getStatisticValues()`.

I don't seem to have access to the support request you raised. Perhaps it's only visible to yourself and the TeamCity team.

We can certainly add the count to the email title. It's also just a template.

**Comment by netwolfuk** [comment]
BTW, I had a look at the code over the weekend. It's very old, and I've learnt a lot since then.
I am refactoring it to not be all global static class variables.

**Comment by netwolfuk** [comment]
Running TeamCity in debug and stepping through the code, I found the following statistics:

- CodeCoverageAbsBTotal
- CodeCoverageAbsRCovered
- CodeCoverageAbsCCovered
- CodeCoverageAbsBCovered
- CodeCoverageAbsMCovered
- CodeCoverageAbsCTotal
- CodeCoverageAbsLCovered
- CodeCoverageAbsLTotal
- CodeCoverageAbsRTotal
- CodeCoverageC
- JacocoComplexityCovered
- CodeCoverageB
- CodeCoverageAbsMTotal
- CodeCoverageL
- JacocoComplexityTotal
- CodeCoverageR
- CodeCoverageM

**Comment by sven-weber** [comment]
Ok, Code Coverage value do not help us regarding code inspection....

**Comment by sven-weber** [comment]
Yeah, maybe the issue is only visible for me and the TC guys.. sorry for that. 
It wasn't that helpful anyways... 

I will have a look at the TC Library documentation to see if there is anything else we might be able to use

**Comment by sven-weber** [comment]
maybe the method getFullStatistics() does help

**Comment by netwolfuk** [comment]
I suspect there will be different items in the Hash of statistics based on the tools.
The actual JavaDoc says it can contain all build messages from the build

> Includes all predefined ValueTypes reported for this build and custom metrics reported by service message

So, to add items to the list, it's possible to use the [ServiceMessage](https://confluence.jetbrains.com/display/TCD18/Build+Script+Interaction+with+TeamCity) function. This is probably how most plugins talk to TeamCity for these metrics.

Probably the best thing for me to do is to copy them all into a statistics object in the template context, and make them available to the template. I can print out debug messages to the teamcity log for every item added to the map.
Then we should be able to do things like...
`$statistics.get('someKey')` inside the template to print out the value.

**Comment by netwolfuk** [comment]
`getFullStatistics()` looks too specific. And I can't see much there other than Test stats.

**Comment by sven-weber** [comment]
You mean to copy all ServiceMessages to the context and make them accessible via in the template?

**Comment by netwolfuk** [comment]
Well sort of. It looks like TeamCity makes them all BigDecimal some how, so it's probably not all of them.

**Comment by sven-weber** [comment]
Ok, its worth a try. 

I took a look at the Log of our FxCop run and it seems to not create ServiceMessages at all and rather publish the results (as xml) via WebDeploy.

```
[14:31:17] : Step 5/10: Run Code Analysis (FxCop) (16s)
[14:31:18] :	 [Step 5/10] Running FxCop
[14:31:18] :	 [Step 5/10] Matched assembly files:
[14:31:18] :	 [Step 5/10]   bin\AnyCPU\Release\Mosaic.exe
[14:31:18] :	 [Step 5/10] Used autodetected FxCop home directory
[14:31:18] :	 [Step 5/10] Starting: "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Team Tools\Static Analysis Tools\FxCop\FxCopCmd.exe" /forceoutput /gac /f:bin\AnyCPU\Release\Mosaic.exe /out:C:\BuildAgent\temp\buildTmp\fxcop-output-2969995639677068277\fxcop-result.xml
[14:31:18] :	 [Step 5/10] in directory: C:\BuildAgent\work\63f3ac10c8f6457d
[14:31:18] :	 [Step 5/10] Microsoft (R) FxCop Command-Line Tool, Version 15.0 (15.0.28010.0) X86
[14:31:18] :	 [Step 5/10] Copyright (C) Microsoft Corporation, All Rights Reserved.
[14:31:18] :	 [Step 5/10] 
[14:31:18] :	 [Step 5/10] Loaded dataflowrules.dll...
[14:31:18] :	 [Step 5/10] Loaded designrules.dll...
[14:31:18] :	 [Step 5/10] Loaded globalizationrules.dll...
[14:31:18] :	 [Step 5/10] Loaded interoperabilityrules.dll...
[14:31:18] :	 [Step 5/10] Loaded maintainabilityrules.dll...
[14:31:18] :	 [Step 5/10] Loaded mobilityrules.dll...
[14:31:18] :	 [Step 5/10] Loaded namingrules.dll...
[14:31:18] :	 [Step 5/10] Loaded performancerules.dll...
[14:31:18] :	 [Step 5/10] Loaded portabilityrules.dll...
[14:31:18] :	 [Step 5/10] Loaded reliabilityrules.dll...
[14:31:18] :	 [Step 5/10] Loaded securityrules.dll...
[14:31:18] :	 [Step 5/10] Loaded securitytransparencyrules.dll...
[14:31:18] :	 [Step 5/10] Loaded usagerules.dll...
[14:31:19] :	 [Step 5/10] Loaded Mosaic.exe...
[14:31:19] :	 [Step 5/10] Initializing Introspection engine...
[14:31:20] :	 [Step 5/10] Analyzing...
[14:31:23] :	 [Step 5/10] Initializing Phoenix engine...
[14:31:23] :	 [Step 5/10] Analyzing...
[14:31:31] :	 [Step 5/10] Analysis Complete.
[14:31:31] :	 [Step 5/10] 
[14:31:31] :	 [Step 5/10] NOTE: One or more referenced assemblies could not be found. Use the '/directory' or '/reference' switch to specify additional assembly reference search paths.
[14:31:31] :	 [Step 5/10] 
[14:31:32] :	 [Step 5/10] Writing 2519 messages...
[14:31:32] :	 [Step 5/10] Writing report to C:\BuildAgent\temp\buildTmp\fxcop-output-2969995639677068277\fxcop-result.xml...
[14:31:32] :	 [Step 5/10] Done:00:00:14.0794078
[14:31:32] :	 [Step 5/10] Process exited with code 0
[14:31:32] :	 [Step 5/10] Publishing artifacts (1s)
[14:31:32] :		 [Publishing artifacts] Collecting files to publish: [C:\BuildAgent\temp\buildTmp\fxcop-output-2969995639677068277\fxcop-result.xml=>.teamcity/FxCop/]
[14:31:33] :		 [Publishing artifacts] Publishing 1 file using [ArtifactsCachePublisher]: C:/BuildAgent/temp/buildTmp/fxcop-output-2969995639677068277/fxcop-result.xml => .teamcity/FxCop
[14:31:33] :		 [Publishing artifacts] Publishing 1 file using [WebPublisher]: C:/BuildAgent/temp/buildTmp/fxcop-output-2969995639677068277/fxcop-result.xml => .teamcity/FxCop
[14:31:32] :	 [Step 5/10] Importing inspection results
[14:31:32] :	 [Step 5/10] Skipped html report generation since not requested
`

This seems kind of strange to me. 

I can check with other runners like Resharper

**Comment by netwolfuk** [comment]
Hmm, perhaps you have a set of [Report Tabs](https://confluence.jetbrains.com/display/TCD18/Including+Third-Party+Reports+in+the+Build+Results) configured.

If that's the case, we probably can't access that data from an `sRunningBuild` instance.
I'll try and get you a build to test in the next hour.

**Comment by netwolfuk** [comment]
Now it prints the following in the `teamcity-server.log` when it prepares the template content.

`DEBUG -   jetbrains.buildServer.SERVER - PrettyEmailBuildStatistics :: Found the following build statistics: CodeCoverageAbsRTotal,buildStageDuration:artifactsPublishing,serverSideBuildFinishing,CodeCoverageAbsLTotal,CodeCoverageAbsLCovered,CodeCoverageAbsCTotal,buildStageDuration:sourcesUpdate,CodeCoverageAbsMCovered,buildStageDuration:buildStepRUNNER_16,JacocoComplexityTotal,buildStageDuration:buildStepRUNNER_20,CodeCoverageAbsBCovered,CodeCoverageAbsMTotal,CodeCoverageAbsCCovered,buildStageDuration:buildFinishing,JacocoComplexityCovered,CodeCoverageAbsRCovered,buildStageDuration:firstStepPreparation,CodeCoverageAbsBTotal,BuildDurationNetTime`

Adding a value to your template looks like this example.
`$info.statistics.getValue('CodeCoverageAbsRTotal')`

**Comment by netwolfuk** [comment]
Build is available [here](https://teamcity.jetbrains.com/viewLog.html?buildId=2005653&buildTypeId=bt159&tab=artifacts) if you have somewhere you can test it.
It would be interesting to see what you get in the log as available statistics keys.

**Comment by sven-weber** [comment]
ok, I will see if I can test it. 

Thank you!

**Comment by sven-weber** [comment]
It took me while but I finally got time to test the plugin. 

The following statistics are found in a build that runs Code Inspection via FxCop: 
```
DEBUG -   jetbrains.buildServer.SERVER - PrettyEmailBuildStatistics :: Found the following
build statistics: 
buildStageDuration:artifactsPublishing,serverSideBuildFinishing,
buildStageDuration:buildFinishing,buildStageDuration:buildStepRUNNER_1,
buildStageDuration:buildStepRUNNER_2,buildStageDuration:sourcesUpdate,
buildStageDuration:firstStepPreparation,BuildDurationNetTime
```
As no Tests where enable the statistics regarding tests are not displayed. 

Unfortunately there seems to be nothing regarding code inspection. 

I also tried adding own statistic value with a CLI buildstep that does the following: 
```
echo ##teamcity[buildStatisticValue key='CodeAnalysis' value='3000']
```

The key is than correctly recognized by the plugin.

**Comment by sven-weber** [comment]
I also took a second look int the TC Documentation and found that there is a class called Inspection Type info (see [here](http://javadoc.jetbrains.net/teamcity/openapi/9.1/jetbrains/buildServer/agent/inspections/InspectionTypeInfo.html)) which seems to have the information we are searching for. 

Unfortunately its located at `jetbrains.buildServer.agent.inspections` and we are mostly working in `jetbrains.buildServer.serverSide` with the plugin. 
I was not able the access the needed information from the object we can work on like SRunningBuild.

**Comment by sven-weber** [comment]
I think we will end up writing a wrapper around the actual inspection tools in order to set the needed statistic values via the service messages. 

Then we can access these values in the Email template and display them. 

Big thanks for your help and the work you put into the plugin!
Really appreciate it

**Comment by netwolfuk** [comment]
This is great @sven-weber. I hoped we might be able to find a simple solution.

With regards to the wrapper, the "teamcity way" is probably to create a "Build Feature" plugin, which we would then add to the build to automatically have the results published as "Service Messages".

There might even already be something like that.

Yes, for the `jetbrains.buildServer.agent` package, these are a type of plugin that runs as part of the build on the agent. I suspect the Build Feature plugin would be an agent plugin of some sort.

**Comment by sven-weber** [comment]
OK thank you I will have a look at Implementing a own Build Feature. 

As currently the Teamcity FxCop, the inspection tool we use, BuildStep is not able to search for the assemblies and you have to enter them all by hand we may end up using a Meta Runner which
- Searches the assemblies
- Runs FxCop
- Parses the Results, sets the Statistics
- Imports the Results for the Teamcity Build Tab

The result import is actually pretty easy using Service Messages
`##teamcity[importData type='FxCop' path='<path to the xml file>']`
this may also work for other inspection tools. 

Thank you again for your great help!

I will start writing the wrapper and then include the statistics in the templates. 

And I will post the results so you may be able to use them too.

**Comment by netwolfuk** [comment]
Yes. Meta runner would certainly be easier. That sounds like a good plan.

Once you have it all running, I'll post your version as a release, or we can work through any other issues and then do a new version.

Thanks for your offer to share your working solution. We can look at conditionally adding any results into the templates using template conditionals. 

And add details on the wiki about how to pass the service messages.

**Comment by sven-weber** [comment]
Sounds like a good plan!

**Comment by sven-weber** [comment]
Hi @netwolfuk, 

I finally had time to test the new version. 

We tried to set Build Statistics with TC Service Messages like the following: 

```
##teamcity[buildStatisticValue key='FxCop_TotalIssues' value='4990']
##teamcity[buildStatisticValue key='FxCop_TotalErrors' value='3166']
##teamcity[buildStatisticValue key='FxCop_TotalWarnings' value='1824']
##teamcity[buildStatisticValue key='InspectionStatsE' value='1000']
```

All Keys are detected by the plugin and we can use them in the email templates:

```
$info.statistics.getValue('FxCop_TotalIssues') TotalIssues, 

$info.statistics.getValue('FxCop_TotalWarnings') TotalWarnings,  

$info.statistics.getValue('FxCop_TotalErrors') TotalErrrors,

$info.statistics.getValue('InspectionStatsE') Test

```

Unfortunately alle values are parsed as doubles rather than integer. 

![qtox_image_2019-03-06 10-13-57 423](https://user-images.githubusercontent.com/39764491/53869221-13509c00-3ff8-11e9-968e-f7c0b0f9de5a.png)

We also tried setting one of the already by TC defined value, like InspectionStatsE which are parsed as integer by TC guaranteed. 

Can we change this to be integer?
Or allow to give a Format when requesting the value.

**Comment by netwolfuk** [comment]
All the values are `BigDecimal`s.
You should be able to call `intValue()` or `intValueExact()` yourself.
eg, `$info.statistics.getValue('FxCop_TotalIssues').intValue()`

However, if that doesn't work, I could add something like `$info.statistics.getIntegerValue('FxCop_TotalIssues')`, which would attempt to call `intValue()` on the value before it is returned.

**Comment by netwolfuk** [comment]
BTW, `$info.statistics.getValue('FxCop_TotalIssues').intValue()` will probably throw a null pointer if the value `FxCop_TotalIssues` is not found. I'm not sure how Velocity handles that.

**Comment by sven-weber** [comment]
Perfect, its working now!

I noticed that a check for null pointers can be performed via: 
```
#if (!$info.statistics.getValue('FxCop_TotalIssues').intValue())
```
([see documentation](https://wiki.apache.org/velocity/CheckingForNull))

If you would like, you can create a release from the current version, everything is working for me. 

I will post the template we used for the Code Inspection later so you can use it if you want. 

Thank you for all the effort and help! 

I really appreciate it!

**Comment by sven-weber** [comment]
the vm file we created for displaying the inspection results looks like the following: 

```
#if (!$info.statistics.getValue('FxCop_TotalIssues').intValue())
<table width=95% class="heading">
        <tr class=heading>
            <td class=heading>Code Inspection</td>
        </tr>
        <tr class="heading">
            <td class="heading">
                <table width=100%>
                    <tr>
                        <td colspan=4>
                            None 
                        </td>
                    </tr>
                </table>
            </td>
        </tr>
    </table>
#else
    <table width=95% class="heading">
        <tr class=heading>
            <td class=heading>Code Inspection</td>
            <td class="heading-link">
             	 <a href="$info.rootURL/viewLog.html?buildId=$info.buildId&buildTypeId=$info.buildTypeId&tab=Inspection">View inspection results</a>
            </td>
        </tr>
        <tr class="heading">
            <td colspan=4 style="background-color:white;">

                            $info.statistics.getValue('FxCop_TotalIssues').intValue() Total Issues

                            <div style="color: #e7d93f">$info.statistics.getValue('FxCop_TotalWarnings').intValue() Warnings</div>  

                            <div style="color: #d00000">$info.statistics.getValue('FxCop_TotalErrors').intValue() Errrors</div>
                            
                </td>
        </tr>
    </table>
#end
```

**Comment by netwolfuk** [comment]
nice work @sven-weber   Thanks for that. I will add the examples to the wiki, and push out a new release.
